### DISCUSSION TITLE
Does anyone already implemented a "midi learn" feature ?

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#393] [Q&A] [answered]
Hi everybody,
I was thinking about a new ctrlr panel whose main purpose would be to translate midi messages.
Let's say CC source is X and CC destination is Y, ctrlr would handle the translation realtime.
To make it easy and convenient It would require a midi learn fonction and a drop-down menu with a list of destination.
Has anyone already implemented a "midi learn" fonction on a ctrlr panel ?

Thanks a lot for your help 

Damien

**Comment by dnaldoog** [comment]
Do you mean a MIDI remapping function?

You can see that in my [Juno panel :)](https://sourceforge.net/projects/juno-alpha/)

**Reply by damiensellier** [reply]
Hi John! Hope everything is alright on your side.
I was thinking about something like you did, yes. But with an extra layer for convenience.
I had the "midi learn" function in mind, like the one you can find on VST plugins.
you select your parameter, click a "learn" button, move a fader/button from a midi controller device, CTRLR catches the incoming midi CC, the "learn" mode switches OFF, then, this midi CC number is stored. 
Now, everytime you move this CC controller from the midi control device, it applies to the "taught/assigned" parameter. 

I bet you know what I'm talking about, it's everywhere on vst synths :)
I really liked the fact you could save the settings of midi CC controls in your JUNO panel, very smart!

**Comment by dnaldoog** [comment, marked as answer]
Hi Damien - it's great to be in contact again - hope you are fine.

The first thing I can think of and it's __not__ tested is something like:

--initialize an empty table & variable when panel loads
midiLearn={} -- you could save this to saveState
globalTmpMod=nil
midiLearnSwitch=false


in a _uiButton_ have a function to change `midiLearnSwitch=false` to `true`

In callback Modulator Function:
<pre>
sendMessage = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
local ccn=26
local n=L(mod:getName())
if  midiLearnSwitch==true then
      globalTmpMod=n -- let's say "myLfo"
      return
elseif midiLearn[n] ~=nil
      panel:sendMidiMessageNow(CtrlrMidiMessage("B0 " ..midiLearn[n].." "..value))
else
-- send the normal number
     panel:sendMidiMessageNow(CtrlrMidiMessage("B0 " ..ccn.." "..value))
end
</pre>

--Next move a slider on the hardware:

--In the midiReceived function you would find the CC number passed in from the hardware:
<pre>
   local s = midi:getSize()
    if s == 3 and midiLearnSwitch ==  true then
    midiLearn[globalTmpMod]=midi:getData():getByte(1) -- assign new midi number to table
    midiLearnSwitch=false -- reset
end
</pre>

Then you would need a way of deleting that table key to return to the default.
Something like that, though it's not really tested!

**Reply by damiensellier** [reply]
Thanks John this is what I was thinking about. 
I'll get deeper into it asap and will propose a sample panel with everything working 100% for anyone looking for this.
A single button "midi learn" followed by a click on a mod or a mod valuechange can initiate the learn function. I'll add a timer to stop the panel waiting to catch the hardware cc change. It will be stopped by a countdown (i.e 10sec) or uppon the cc message sent by the hardware like you proposed.

That's a really nice feature! Can't wait to work on this.

Have a nice day ;)
Damien

**Reply by dnaldoog** [reply]
Look forward to seeing how it goes!
